### PR TITLE
Check for rustdocs build warnings in CI

### LINKED ISSUE #1403: Bunch of doc warnings
`cargo +nightly doc` shows a bunch of warnings. We should fix them and add a check to CI to avoid them in the future.

### PR DESCRIPTION
**Pull request by tcharding**
Currently we do not fail the CI script if the docs build throws warnings, since we are a group of super anal, easily triggered, code cleanliness obsessed devs this causes a mild rash to develop on the lower back [0]. We can easily fix this by checking for build warnings in CI.

[0] - Amusingly my rash has been playing up since Friday but I thought I'd fixed the warnings in an open PR someplace so I was ignoring it, seeing Kixunil's [issue](https://github.com/rust-bitcoin/rust-bitcoin/issues/1403) this morning prompted me to fix it :)

Fix #1403

### CODE REVIEW

**Review by Kixunil — DISMISSED** [review]
ACK 8847475e636bf536c2c6083634b505d1eb75be59

If CI succeeds.

**Comment by tcharding** [comment]
Agh, had to use `--all-features` to get the docs to build, without it (using explicit feature list) we get
```
error: too many file operands
```

**Comment by tcharding** [comment]
I had used ` <bash command> | exit 1` instead of `<bash command> || exit 1` :)

**Comment by Kixunil** [comment]
I think `--features` requires them to be comma-separated.

> I had used `<bash command> | exit 1` instead of `<bash command> || exit 1` :)

Lol, that was quite a confusing.

**Comment by tcharding** [comment]
> I think --features requires them to be comma-separated.

Yeah I cannot work out how the variable interpolation is working in the CI script, there is another example of it that works when it looks to me like it shouldn't
```
cargo build --verbose --features="no-std $FEATURES" --no-default-features
```
On the command line this has to use comma separation as you say?

**Review by Kixunil — APPROVED** [review]
ACK 108a1f73ca8bc14e7b3bed940840b18b2e8b0771

**Comment by Kixunil** [comment]
No clue about that, I'm going to sleep now. :D

**Review by apoelstra — APPROVED** [review]
ACK 108a1f73ca8bc14e7b3bed940840b18b2e8b0771

**Comment by Kixunil** [comment]
This fix either doesn't work or broke again at some point. There are doc warnings again.

File: bitcoin/contrib/test.sh [hunk 1]
```diff
@@ -81,7 +81,7 @@ cargo run --example taproot-psbt --features=bitcoinconsensus
 
 # Build the docs if told to (this only works with the nightly toolchain)
 if [ "$DO_DOCS" = true ]; then
-    RUSTDOCFLAGS="--cfg docsrs" cargo +nightly rustdoc --features="$FEATURES" -- -D rustdoc::broken-intra-doc-links
+    RUSTDOCFLAGS="--cfg docsrs" cargo +nightly rustdoc --all-features -- -D rustdoc::broken-intra-doc-links -D warnings || exit 1
```
Comment (tcharding) [inline bitcoin/contrib/test.sh:84]: We have `set -ex` so this "fix" did not add any new behaviour to the script :(

Comment (Kixunil) [reply bitcoin/contrib/test.sh:84]: I think `-D` doesn't work for some reason.

**Comment by tcharding** [comment]
> This fix either doesn't work or broke again at some point. There are doc warnings again.

Are you still seeing warnings?

I'm on "comit: 46eda5fc Merge rust-bitcoin/rust-bitcoin#1499: Add `from_bytes(Vec<u8>)` to `ScriptBuf`"

And `RUSTDOCFLAGS="--cfg docsrs" cargo +nightly rustdoc --all-features -- -D rustdoc::broken-intra-doc-links` shows no warnings, what command are you using?

**Comment by Kixunil** [comment]
Commit 46eda5fc332f7db5a71829717789df7509f5a6fc
Command: `cargo +nightly doc --open`
Output:

```
warning: unresolved link to `absolute::LockTime`
   --> bitcoin/src/blockdata/locktime/absolute.rs:211:11
    |
211 |     /// [`absolute::LockTime`].
    |           ^^^^^^^^^^^^^^^^^^ no item named `absolute` in scope
    |
    = note: `#[warn(rustdoc::broken_intra_doc_links)]` on by default

warning: unresolved link to `relative::LockTime`
  --> bitcoin/src/blockdata/locktime/relative.rs:37:32
   |
37 |     /// Returns true if this [`relative::LockTime`] is satisfied by either height or time.
   |                                ^^^^^^^^^^^^^^^^^^ no item named `relative` in scope

warning: unresolved link to `relative::LockTime`
  --> bitcoin/src/blockdata/locktime/relative.rs:66:11
   |
66 |     /// [`relative::LockTime`].
   |           ^^^^^^^^^^^^^^^^^^ no item named `relative` in scope

warning: unresolved link to `relative::LockTime`
   --> bitcoin/src/blockdata/locktime/relative.rs:105:32
    |
105 |     /// Returns true if this [`relative::LockTime`] is satisfied by [`Height`].
    |                                ^^^^^^^^^^^^^^^^^^ no item named `relative` in scope

warning: unresolved link to `relative::LockTime`
   --> bitcoin/src/blockdata/locktime/relative.rs:131:32
    |
131 |     /// Returns true if this [`relative::LockTime`] is satisfied by [`Time`].
    |                                ^^^^^^^^^^^^^^^^^^ no item named `relative` in scope

warning: `bitcoin` (lib doc) generated 5 warnings
```

But no warnings with the `docsrs` flag. WTF.

**Comment by Kixunil** [comment]
Oh, found it.

**Comment by Kixunil** [comment]
See #1504

**Pull request merged**